tools/setup_helpers/ninja_builder.py

```diff
@@ -0,0 +1,81 @@
+import os
+import setuptools
+import distutils
+from contextlib import contextmanager
+
+
+# on the fly create a ninja file in build/ and then
+# run it when run() is called.
+class NinjaBuilder(object):
+    def __init__(self, name):
+        import ninja
+        build_dir = 'build'
+        if not os.path.exists(build_dir):
+            os.mkdir(build_dir)
+        self.name = name
+        self.filename = os.path.join(build_dir, 'build.{}.ninja'.format(name))
+        print('Ninja build file at {}'.format(self.filename))
+        self.writer = ninja.Writer(open(self.filename, 'w'))
+        self.writer.rule('do_cmd', '$cmd')
+
+    def run(self):
+        import ninja
+        self.writer.close()
+        assert(0 == ninja._program('ninja', ['-f', self.filename]))
```

Comment:
What is this line doing? It doesn’t have any side effects, right?

If this is the command that really launches the build, then it would be a lot better to use an `if` to check for errors. Assertions can be disabled from the command line, and this would make our build a no-op.

Reply:
It seems to have side effects; `ninja -f filename` invokes the build described in filename for default target.

Reply:
I switched it to subprocess.check_call to make it more clear.

Reply:
It's better to use [`ninja._program`](https://github.com/scikit-build/ninja-python-distributions/blob/master/ninja/__init__.py#L32). The ninja python package has bundled together a ninja binary, and this function will find it. Directly using `subprocess.check_call` will needs user to have ninja in their `$PATH`.